### PR TITLE
Add nonce to tx

### LINKED ISSUE #80: Add nonce to transactions
When we have an error in submitting a graph update to the chain, we will re-process the failed job. If the re-process attempt results in the same extrinsic payload as the initial failed attempt, it may result in a failure due to the identical transaction already being present in the pool.

Suggested solution: we should set a nonce when submitting the graph update transaction

### PR DESCRIPTION
**Pull request by saraswatpuneet**
## Details
We would want to send txs with nonce to prevent errors pertaining to stale/duplicate txs

* Add a nonce service (copied from content publishing)
* Pass nonce to `signAndSend`

Tests:
- [x] run a full migration test

Closes #80

### CODE REVIEW

File: src/processor/reconnection-graph.service.ts [hunk 1]
```diff
@@ -434,9 +436,10 @@ export class ReconnectionGraphService {
     this.logger.debug(`Submitting batch for user ${dsnpUserId.toString()}, batch length: ${batch.length}, batch: ${JSON.stringify(batch)}`);
     try {
       const currrentEpoch = await this.blockchainService.getCurrentCapacityEpoch();
+      const nonce = await this.nonceService.getNextNonce();
       const [event, eventMap] = await this.blockchainService
         .createExtrinsic({ pallet: 'frequencyTxPayment', extrinsic: 'payWithCapacityBatchAll' }, { eventPallet: 'utility', event: 'BatchCompleted' }, providerKeys, batch)
-        .signAndSend();
+        .signAndSend(nonce);
```
Comment (JoeCap08055) [inline src/processor/reconnection-graph.service.ts:442]: Should this be:
```suggestion
        .signAndSend({ nonce });
```

File: src/blockchain/blockchain.service.ts [hunk 2]
```diff
@@ -135,4 +135,8 @@ export class BlockchainService implements OnApplicationBootstrap, OnApplicationS
     const epochLength: u32 = await this.query('capacity', 'epochLength');
     return typeof epochLength === 'number' ? epochLength : epochLength.toNumber();
   }
+
+  public async getNonce(account: Uint8Array): Promise<number> {
+    return this.rpc('system', 'accountNextIndex', account);
+  }
```
Comment (JoeCap08055) [inline src/blockchain/blockchain.service.ts:141]: I think `accountNextIndex` returns an `Index`, which is a `u32`, not a JS `number`

Comment (saraswatpuneet) [reply src/blockchain/blockchain.service.ts:141]: I think the wrappers internally be calling `toNumber` before returning, assuming we would never reach a bigint


Do we want to make it bigint?

File: src/processor/nonce.service.ts [hunk 3]
```diff
@@ -0,0 +1,57 @@
...
+    return nextNonce;
+  }
+
+  // eslint-disable-next-line class-methods-use-this
+  getNextPossibleKeys(currentNonce: number): string[] {
+    const keys: string[] = [];
+    for (let i = 0; i < RedisUtils.NUMBER_OF_NONCE_KEYS_TO_CHECK; i += 1) {
+      const key = currentNonce + i;
+      keys.push(RedisUtils.getNonceKey(`${key}`));
+    }
```
Comment (JoeCap08055) [inline src/processor/nonce.service.ts:54]: code golf (nit):
```suggestion
    const keys: string[] = Array(RedisUtils.NUMBER_OF_NONCE_KEYS_TO_CHECK).fill(null).map((_entry, i) => RedisUtils.getNonceKey(`${currentNonce + i}`));
```

**Comment by saraswatpuneet** [comment]
All tests have passed and no errors were seen that were seen before Nonce got introduced.

**Review by JoeCap08055 — APPROVED** [review]

**Pull request merged**